src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -107,32 +117,278 @@ internal override IEnumerable<TSource> GetEnumerable(int index)
                 }
             }
         }
 
+        private sealed class Concat2CollectionIterator<TSource> : ConcatCollectionIterator<TSource>
+        {
+            private readonly ICollection<TSource> _first;
+            private readonly ICollection<TSource> _second;
+
+            internal Concat2CollectionIterator(ICollection<TSource> first, ICollection<TSource> second)
+            {
+                Debug.Assert(first != null && second != null);
+                _first = first;
+                _second = second;
+            }
+
+            internal override int Count => checked(_first.Count + _second.Count);
+
+            public override Iterator<TSource> Clone()
+            {
+                return new Concat2CollectionIterator<TSource>(_first, _second);
+            }
+
+            internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
+            {
+                var nextCol = next as ICollection<TSource>;
+                if (nextCol != null)
+                {
+                    return new ConcatNCollectionIterator<TSource>(this, nextCol, 2);
+                }
+                return new ConcatNEnumerableIterator<TSource>(this, next, 2);
+            }
+
+            internal override void CopyToUntil(TSource[] array, int lastIndex)
+            {
+                Debug.Assert(array != null && array.Length != 0);
+                Debug.Assert(lastIndex >= 0 && lastIndex < array.Length);
+
+                int totalCount = lastIndex + 1;
+                Debug.Assert(totalCount >= Count);
+
+                int firstCount = _first.Count;
+                int secondCount = _second.Count;
+
+                _second.CopyTo(array, totalCount - secondCount);
+                _first.CopyTo(array, totalCount - secondCount - firstCount);
+            }
+
+            internal override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                switch (index)
+                {
+                    case 0: return _first;
+                    case 1: return _second;
+                    default: return null;
+                }
+            }
+
+            public override TSource[] ToArray()
+            {
+                int firstCount = _first.Count;
+                int secondCount = _second.Count;
+                int totalCount = firstCount + secondCount;
+
+                if (totalCount == 0)
+                {
+                    return Array.Empty<TSource>();
+                }
+
+                var result = new TSource[totalCount];
+
+                _first.CopyTo(result, 0);
+                _second.CopyTo(result, firstCount);
+
+                return result;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int firstCount = _first.Count;
+                int secondCount = _second.Count;
+                var result = new List<TSource>(firstCount + secondCount);
+                
+                if (firstCount != 0)
```

Comment:
Do these conditionals on the length add much value?  If not, I'd just assume not have them and let AddRange do the right thing, just as was done with CopyTo in ToArray.